File: app/api/aica_api/services/event_plan.py
"""Event plan freeze service (T015) — deterministic per-tick schedule.

Pure, side-effect-free. Same scenario → same EventPlan every call.

For each tick index i in 0..N-1 (N = total_duration // tick_seconds):
  - route_fraction = min(1.0, i * tick_seconds / total_duration)
  - drowsiness_band: step-held from event_presets.drowsiness_schedule
  - signal_duration: step-held from signal_duration_schedule (extra field) or
    fallback to signal_duration_at_trigger for all ticks
  - rest_spot_eta: 'near' before the rest_spot_eta_near_before segment's at
    position, 'none' at or after; or from rest_spot_eta_schedule if present;
    defaults to 'none' if neither is configured
  - continuous_driving_time: ordinal band from binning(elapsed_seconds)
  - active_segment_id: step-held from route_intent.segments by route_fraction
"""

from __future__ import annotations

from aica_api.models.run import EventPlan, TickPlanEntry
from aica_api.models.scenario import ScenarioDef
from aica_api.services.binning import bin_context


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _step_hold(
    schedule: list[dict],
    fraction: float,
    at_key: str = "at",
    val_key: str = "band",
    default: str = "",
) -> str:
    """Step-held schedule lookup: last entry whose at_key value ≤ fraction."""
    result = default
    for entry in schedule:
        if entry[at_key] <= fraction:
            result = entry[val_key]
    return result


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def freeze_event_plan(scenario: ScenarioDef) -> EventPlan:
    """Freeze a deterministic per-tick event plan for a run.

    This is the only place where the scenario's event_presets are resolved
    into a per-tick schedule.  The resulting EventPlan is stored in the run
    log and may not be mutated after the run starts.

    Args:
        scenario: A validated ScenarioDef.

    Returns:
        A fully-populated EventPlan with N = total_duration // tick_seconds
        TickPlanEntry instances, one per simulation tick.
    """
    total_duration = scenario.total_duration_seconds
    tick_seconds = scenario.tick_seconds
    n_ticks = total_duration // tick_seconds

    # ── Drowsiness schedule ────────────────────────────────────────────────
    drowsiness_schedule = [
        {"at": e.at, "band": e.band}
        for e in scenario.event_presets.drowsiness_schedule
    ]

    # ── Signal duration schedule (optional extra field) ────────────────────
    extra = scenario.event_presets.model_extra or {}
    signal_duration_schedule: list[dict] | None = extra.get("signal_duration_schedule")
    signal_duration_at_trigger: str = scenario.event_presets.signal_duration_at_trigger

    # ── Rest spot ETA ──────────────────────────────────────────────────────
    rest_spot_eta_near_before: str | None = scenario.event_presets.rest_spot_eta_near_before
    rest_spot_eta_schedule: list[dict] | None = scenario.event_presets.rest_spot_eta_schedule

    # Find the rest facility's at-fraction when using near_before logic
    rest_facility_at: float | None = None
    if rest_spot_eta_near_before is not None:
        for seg in scenario.route_intent.segments:
            if seg.id == rest_spot_eta_near_before:
                rest_facility_at = seg.at
                break

    # ── Segments for active_segment_id lookup ─────────────────────────────
    segments = scenario.route_intent.segments

    # ── Build ticks ───────────────────────────────────────────────────────
    ticks: list[TickPlanEntry] = []
    for i in range(n_ticks):
        elapsed_seconds = i * tick_seconds
        route_fraction = min(1.0, elapsed_seconds / total_duration)

        # Drowsiness band (step-held)
        drowsiness_band = _step_hold(drowsiness_schedule, route_fraction) or "none"

        # Signal duration
        if signal_duration_schedule:
            signal_duration = (
                _step_hold(signal_duration_schedule, route_fraction)
                or "transient"
            )
        else:
            # Fallback: use signal_duration_at_trigger for all ticks
            signal_duration = signal_duration_at_trigger

        # Rest spot ETA
        if rest_spot_eta_schedule:
            rest_spot_eta = (
                _step_hold(rest_spot_eta_schedule, route_fraction) or "none"
            )
        elif rest_facility_at is not None:
            # near before the facility, none at/after
            rest_spot_eta = "near" if route_fraction < rest_facility_at else "none"
        else:
            rest_spot_eta = "none"

        # Continuous driving time via boundary-binning
        banded = bin_context({"travel_time_sec": elapsed_seconds})
        continuous_driving_time = banded["continuous_driving_time"]

        # Active segment (step-held by at)
        active_segment_id = segments[0].id
        for seg in segments:
            if seg.at <= route_fraction:
                active_segment_id = seg.id

        entry = TickPlanEntry(
            tick_index=i,
            drowsiness_band=drowsiness_band,
            route_fraction=route_fraction,
            signal_duration=signal_duration,
            rest_spot_eta=rest_spot_eta,
            # Extra fields (model_config extra="allow"):
            elapsed_seconds=elapsed_seconds,
            continuous_driving_time=continuous_driving_time,
            active_segment_id=active_segment_id,
        )
        ticks.append(entry)

    return EventPlan(ticks=ticks)

File: htmlapp/frontend/src/engine/event_plan.ts
/**
 * Event plan freeze service.
 *
 * Ported from `app/api/aica_api/services/event_plan.py` (behavior-of-record).
 * Pure, side-effect-free: same scenario -> same EventPlan every call.
 *
 * `freeze_event_plan` (M1) resolves a scenario's event_presets into a
 * per-tick schedule (ticks[]). `build_event_plan` (M2) builds the
 * event-level schedule (tick_seconds + traffic/weather/rest event lists)
 * consumed by the M2 tick engine; it does NOT populate ticks[].
 *
 * Only the exported function identifiers (freezeEventPlan, buildEventPlan)
 * are camelCased. All object keys and every band/event string value are
 * preserved byte-for-byte from the Python (snake_case) because they cross
 * the parity boundary.
 *
 * `freezeEventPlan` takes an optional second `seed` parameter that is
 * IGNORED: the Python `freeze_event_plan(scenario)` has no seed argument at
 * all (the engine is fully deterministic without RNG — see prng.ts). The
 * parameter exists only so this function's call signature matches the test
 * harness's `freezeEventPlan(fx.input.scenario, fx.input.seed)`.
 */

import type { RouteFacts, RouteSegment, ScenarioDef } from '../api/types'
import { binContext } from './binning'

// ---------------------------------------------------------------------------
// Public types (mirror aica_api.models.run — genuinely absent from api/types.ts)
// ---------------------------------------------------------------------------

export type TickPlanEntry = {
  tick_index: number
  drowsiness_band: string
  route_fraction: number
  signal_duration: string
  rest_spot_eta: string
  // Extra fields (Python model_config extra="allow"); always set by freezeEventPlan.
  elapsed_seconds: number
  continuous_driving_time: string
  active_segment_id: string
}

export type TrafficEvent = {
  id: string
  start_min: number
  duration_min: number
  affected_segment_id: string
  speed_kph: number
  // Optional POSITION-native km bounds (fixbug-0806). When present, the tick
  // engine gates congestion on distance_km instead of elapsed_min — the
  // correct axis, since routes are not time-linear in distance once segment
  // speeds and auto-rest stops are involved. See `activeTrafficJam` in
  // tick_engine.ts.
  start_km?: number | null
  end_km?: number | null
}

export type WeatherEvent = {
  id: string
  start_min: number
  duration_min: number
}

export type RestOpportunity = {
  id: string
  route_position_km: number
}

export type EventPlan = {
  // M1 — per-tick plan (populated by freezeEventPlan)
  ticks: TickPlanEntry[]
  // M2 — event-level schedule (populated by buildEventPlan)
  tick_seconds: number
  traffic_events: TrafficEvent[]
  weather_events: WeatherEvent[]
  rest_opportunities: RestOpportunity[]
  // Feature 009 — frozen run seed for the anomaly generator (Principle III).
  run_seed: number
}

/** Pydantic EventPlan() field defaults — used for error-path drafts in run_plan.ts. */
export function defaultEventPlan(): EventPlan {
  return { ticks: [], tick_seconds: 60, traffic_events: [], weather_events: [], rest_opportunities: [], run_seed: 42 }
}

/**
 * ScenarioDef fields consumed by the M2 engine (build_event_plan, tick_engine,
 * run_plan) that are genuinely absent from — or mistyped in — the synced
 * `../api/types.ts` ScenarioDef (which lags the Python model: no `presets`,
 * no `is_night`; `initial_state` is typed `Record<string,string>` but actual
 * scenario JSON carries numeric values). Declared here as a local extension;
 * `types.ts` itself is never edited. Both bundled scenarios always carry
 * `presets` and `is_night` (Python field defaults: `presets={}`, `is_night=False`).
 */
export type ScenarioDefM2 = Omit<ScenarioDef, 'initial_state'> & {
  presets: Record<string, unknown>
  is_night: boolean
  rest_drowsiness_ceiling?: number
  initial_state: Record<string, unknown>
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

type ScheduleEntry = { at: number; band: string }

/** Step-held schedule lookup: last entry whose at_key value <= fraction. */
function stepHold(
  schedule: ScheduleEntry[],
  fraction: number,
  atKey: keyof ScheduleEntry = 'at',
  valKey: keyof ScheduleEntry = 'band',
  defaultVal = '',
): string {
  let result: string = defaultVal
  for (const entry of schedule) {
    if (entry[atKey] as unknown as number <= fraction) {
      result = entry[valKey] as unknown as string
    }
  }
  return result
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Freeze a deterministic per-tick event plan for a run (M1).
 *
 * `seed` is IGNORED — see module docstring. Present only for test-harness
 * call-signature compatibility.
 */
export function freezeEventPlan(scenario: ScenarioDefM2, _seed?: number): EventPlan {
  const totalDuration = scenario.total_duration_seconds
  const tickSeconds = scenario.tick_seconds
  const nTicks = Math.floor(totalDuration / tickSeconds)

  // event_presets extra fields (Python model_config extra="allow"): not
  // declared on the synced EventPreset type, read via a loose cast.
  const eventPresetsExtra = scenario.event_presets as unknown as Record<string, unknown>
  const drowsinessSchedule = (eventPresetsExtra['drowsiness_schedule'] as ScheduleEntry[] | undefined) ?? []
  const signalDurationSchedule = eventPresetsExtra['signal_duration_schedule'] as ScheduleEntry[] | undefined
  const signalDurationAtTrigger: string = scenario.event_presets.signal_duration_at_trigger

  const restSpotEtaNearBefore: string | null = scenario.event_presets.rest_spot_eta_near_before ?? null
  const restSpotEtaSchedule = scenario.event_presets.rest_spot_eta_schedule as ScheduleEntry[] | undefined

  // Find the rest facility's at-fraction when using near_before logic.
  let restFacilityAt: number | null = null
  if (restSpotEtaNearBefore !== null) {
    for (const seg of scenario.route_intent.segments) {
      if (seg.id === restSpotEtaNearBefore) {
        restFacilityAt = seg.at
        break
      }
    }
  }

  const segments: RouteSegment[] = scenario.route_intent.segments

  const ticks: TickPlanEntry[] = []
  for (let i = 0; i < nTicks; i++) {
    const elapsedSeconds = i * tickSeconds
    const routeFraction = Math.min(1.0, elapsedSeconds / totalDuration)

    // Drowsiness band (step-held)
    const drowsinessBand = stepHold(drowsinessSchedule, routeFraction) || 'none'

    // Signal duration
    let signalDuration: string
    if (signalDurationSchedule && signalDurationSchedule.length > 0) {
      signalDuration = stepHold(signalDurationSchedule, routeFraction) || 'transient'
    } else {
      // Fallback: use signal_duration_at_trigger for all ticks
      signalDuration = signalDurationAtTrigger
    }

    // Rest spot ETA
    let restSpotEta: string
    if (restSpotEtaSchedule && restSpotEtaSchedule.length > 0) {
      restSpotEta = stepHold(restSpotEtaSchedule, routeFraction) || 'none'
    } else if (restFacilityAt !== null) {
      // near before the facility, none at/after
      restSpotEta = routeFraction < restFacilityAt ? 'near' : 'none'
    } else {
      restSpotEta = 'none'
    }

    // Continuous driving time via boundary-binning
    const continuousDrivingTime = driveTimeBand(elapsedSeconds)

    // Active segment (step-held by at)
    let activeSegmentId = segments[0].id
    for (const seg of segments) {
      if (seg.at <= routeFraction) {
        activeSegmentId = seg.id
      }
    }

    ticks.push({
      tick_index: i,
      drowsiness_band: drowsinessBand,
      route_fraction: routeFraction,
      signal_duration: signalDuration,
      rest_spot_eta: restSpotEta,
      elapsed_seconds: elapsedSeconds,
      continuous_driving_time: continuousDrivingTime,
      active_segment_id: activeSegmentId,
    })
  }

  // Python: EventPlan(ticks=ticks) — every other field takes its declared
  // model default (tick_seconds=60, empty event lists), regardless of the
  // scenario's own tick_seconds. Reproduced verbatim, not "fixed".
  return { ticks, tick_seconds: 60, traffic_events: [], weather_events: [], rest_opportunities: [], run_seed: 42 }
}

/**
 * Build the M2 EventPlan from RouteFacts and scenario presets.
 *
 * Returns an EventPlan with M2 fields (tick_seconds, traffic_events,
 * weather_events, rest_opportunities) and NO per-tick ticks[] entries.
 *
 * Caller-supplied `presets` are merged on top of `scenario.presets`
 * (caller wins on key conflict).
 */
export function buildEventPlan(
  routeFacts: RouteFacts,
  scenario: ScenarioDefM2,
  presets?: Record<string, unknown> | null,
  runSeed?: number | null,
): EventPlan {
  const mergedPresets: Record<string, unknown> = { ...(scenario.presets ?? {}) }
  if (presets) {
    Object.assign(mergedPresets, presets)
  }

  const tickSeconds = Number(mergedPresets['tick_seconds'] ?? scenario.tick_seconds)

  const trafficEvents: TrafficEvent[] = ((mergedPresets['traffic_events'] as Record<string, unknown>[] | undefined) ?? [])
    .map((raw) => raw as unknown as TrafficEvent)

  const weatherEvents: WeatherEvent[] = ((mergedPresets['weather_events'] as Record<string, unknown>[] | undefined) ?? [])
    .map((raw) => raw as unknown as WeatherEvent)

  const restOpportunities: RestOpportunity[] = routeFacts.rest_spot_positions.map((posKm, i) => ({
    id: `rest_${i}`,
    route_position_km: posKm,
  }))

  const frozenRunSeed = runSeed != null
    ? runSeed
    : (((scenario as unknown as Record<string, unknown>)['run_seed_default'] as number | undefined) ?? 42)

  return {
    ticks: [],
    tick_seconds: tickSeconds,
    traffic_events: trafficEvents,
    weather_events: weatherEvents,
    rest_opportunities: restOpportunities,
    run_seed: frozenRunSeed,
  }
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

/** Convert elapsed seconds to a continuous_driving_time band via binning. */
function driveTimeBand(elapsedSeconds: number): string {
  const banded = binContext({ travel_time_sec: elapsedSeconds })
  return banded['continuous_driving_time'] as string
}
